pub/remote_forward.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	log "github.com/inconshreveable/log15"
+	"github.com/pkg/errors"
+)
+
+func remoteForwardEndpoint(c *gin.Context, cfg *Config, jobPath string) {
+	requestId := getRequestTracingId(c.Request, cfg.RequestTracingHeader)
+	logger := log.New(log.Ctx{
+		"requestId": requestId,
+	})
+
+	logger.Info("Incoming forwarding request from main Pub", log.Ctx{"method": c.Request.Method, "path": c.Request.URL.Path})
+	statusCode, err := handleRemoteForwardRequest(c, cfg, logger, requestId, jobPath)
+	if err != nil {
+		errorStr := err.Error()
+		logger.Error("Proxy request error", log.Ctx{
+			"status": statusCode,
+			"error":  errorStr,
+			"path":   c.Request.URL.Path,
+		})
+		c.JSON(statusCode, gin.H{
+			"error":     fmt.Sprintf("Proxy request error: %s", errorStr),
+			"status":    http.StatusText(statusCode),
+			"requestId": requestId,
+		})
+	}
+}
+
+func handleRemoteForwardRequest(
+	c *gin.Context,
+	cfg *Config,
+	logger log.Logger,
+	requestId string,
+	jobPath string,
+) (int, error) {
+	if !cfg.RemoteGatewayMode {
+		return http.StatusUnauthorized, errors.New("Forwarding endpoint is only available in remote gateway mode")
+	}
+
+	if c.Request.Method != "POST" && c.Request.Method != "GET" {
+		c.Writer.Header().Set("Allow", "GET, POST")
+		return http.StatusMethodNotAllowed, errors.New("Method not allowed")
+	}
+
+	jobName := c.Param("job")
+	if jobName == "" {
+		return http.StatusBadRequest, errors.New("Couldn't extract job name")
+	}
+	jobVersion := c.Param("version")
+	if jobVersion == "" {
```

Comment:
Indeed, it's duplicated, but because of this `return` statement inside, there's not much flexibility in moving it to a function. The best I could do is:
```go
jobVersion, err := extractParam(c, "version", "job version")
if err != nil {
    return http.StatusBadRequest, err
}
```
But this doesn't reduce number of lines. On the contrary, there's additional function. Therefore, I'm inclined to leaving it in its current form.

Reply:
Yeah, that's a fair assessment. Resolving.